File: src/tools/check-status.ts
import { exec } from "child_process";
import { promisify } from "util";

const execAsync = promisify(exec);

interface CheckClusterStatusArgs {
  namespace?: string;
  releaseName?: string;
  detailed?: boolean;
}

export async function checkClusterStatus(args: any) {
  const {
    namespace,
    releaseName,
    detailed = false,
  } = args as CheckClusterStatusArgs;

  try {
    let response = `# Cluster Status Check\n\n`;

    // Check cluster connectivity
    response += `## Cluster Connectivity\n\n`;
    try {
      const { stdout } = await execAsync("oc cluster-info");
      response += `✅ Connected to cluster\n\n\`\`\`\n${stdout}\`\`\`\n\n`;
    } catch (error) {
      response += `❌ Not connected to cluster\n\n`;
      return {
        content: [{ type: "text", text: response }],
        isError: true,
      };
    }

    // Check namespace-specific resources
    if (namespace) {
      response += `## Namespace: ${namespace}\n\n`;
      
      // Check pods
      try {
        let podCommand = `oc get pods -n ${namespace}`;
        if (releaseName) {
          podCommand += ` -l app.kubernetes.io/instance=${releaseName}`;
        }
        const { stdout: podStatus } = await execAsync(podCommand);
        response += `### Pods\n\n\`\`\`\n${podStatus}\`\`\`\n\n`;
      } catch (error) {
        response += `### Pods\n\n❌ Unable to retrieve pods\n\n`;
      }

      // Check services
      if (releaseName) {
        try {
          const { stdout: svcStatus } = await execAsync(
            `oc get svc -n ${namespace} -l app.kubernetes.io/instance=${releaseName}`
          );
          response += `### Services\n\n\`\`\`\n${svcStatus}\`\`\`\n\n`;
        } catch (error) {
          response += `### Services\n\n❌ Unable to retrieve services\n\n`;
        }
      }

      // Check routes
      if (releaseName) {
        try {
          const { stdout: routeStatus } = await execAsync(
            `oc get route -n ${namespace}`
          );
          response += `### Routes\n\n\`\`\`\n${routeStatus}\`\`\`\n\n`;
        } catch (error) {
          response += `### Routes\n\n❌ No routes found\n\n`;
        }
      }

      // Check Vault status if specified
      if (releaseName && detailed) {
        response += `### Vault Status\n\n`;
        try {
          const { stdout: vaultStatus } = await execAsync(
            `oc exec -n ${namespace} ${releaseName}-0 -- vault status`
          );
          response += `\`\`\`\n${vaultStatus}\`\`\`\n\n`;
        } catch (error) {
          response += `❌ Unable to check vault status\n\n`;
        }
      }

      // Check Helm releases
      if (releaseName) {
        response += `### Helm Release\n\n`;
        try {
          const { stdout: helmStatus } = await execAsync(
            `helm list -n ${namespace} -f ${releaseName}`
          );
          response += `\`\`\`\n${helmStatus}\`\`\`\n\n`;
        } catch (error) {
          response += `❌ Unable to retrieve helm release info\n\n`;
        }
      }
    }

    // Overall health summary
    response += `## Health Summary\n\n`;
    response += await generateHealthSummary(namespace, releaseName);

    return {
      content: [{ type: "text", text: response }],
    };
  } catch (error) {
    const errorMessage = error instanceof Error ? error.message : String(error);
    return {
      content: [
        {
          type: "text",
          text: `Status check failed: ${errorMessage}`,
        },
      ],
      isError: true,
    };
  }
}

async function generateHealthSummary(namespace?: string, releaseName?: string) {
  let summary = ``;
  
  if (!namespace) {
    summary += `✅ Cluster is accessible\n`;
    summary += `ℹ️ Specify namespace for detailed health check\n`;
    return summary;
  }

  try {
    const { stdout } = await execAsync(
      `oc get pods -n ${namespace} ${releaseName ? `-l app.kubernetes.io/instance=${releaseName}` : ""} -o json`
    );
    const pods = JSON.parse(stdout);
    
    if (!pods.items || pods.items.length === 0) {
      summary += `⚠️ No pods found\n`;
      return summary;
    }

    const totalPods = pods.items.length;
    const runningPods = pods.items.filter((p: any) => p.status.phase === "Running").length;
    const readyPods = pods.items.filter((p: any) => 
      p.status.conditions?.some((c: any) => c.type === "Ready" && c.status === "True")
    ).length;

    summary += `**Pods:** ${runningPods}/${totalPods} running, ${readyPods}/${totalPods} ready\n`;
    
    if (runningPods === totalPods && readyPods === totalPods) {
      summary += `✅ All pods are healthy\n`;
    } else {
      summary += `⚠️ Some pods are not healthy\n`;
    }
  } catch (error) {
    summary += `❌ Unable to generate health summary\n`;
  }

  return summary;
}

File: python_src/tools/check_status.py
from __future__ import annotations

import json
from typing import Any

from .common import run_shell, text_response, ToolResponse


def check_cluster_status(args: dict[str, Any]) -> ToolResponse:
    namespace = args.get("namespace")
    release_name = args.get("releaseName")
    detailed = bool(args.get("detailed", False))

    try:
        response = "# Cluster Status Check\n\n"

        response += "## Cluster Connectivity\n\n"
        try:
            stdout, _ = run_shell("oc cluster-info")
            response += f"Connected to cluster\n\n```\n{stdout}```\n\n"
        except Exception:
            response += "Not connected to cluster\n\n"
            return text_response(response, is_error=True)

        if namespace:
            response += f"## Namespace: {namespace}\n\n"

            try:
                pod_command = f"oc get pods -n {namespace}"
                if release_name:
                    pod_command += f" -l app.kubernetes.io/instance={release_name}"
                pod_status, _ = run_shell(pod_command)
                response += f"### Pods\n\n```\n{pod_status}```\n\n"
            except Exception:
                response += "### Pods\n\nUnable to retrieve pods\n\n"

            if release_name:
                try:
                    svc_status, _ = run_shell(f"oc get svc -n {namespace} -l app.kubernetes.io/instance={release_name}")
                    response += f"### Services\n\n```\n{svc_status}```\n\n"
                except Exception:
                    response += "### Services\n\nUnable to retrieve services\n\n"

                try:
                    route_status, _ = run_shell(f"oc get route -n {namespace}")
                    response += f"### Routes\n\n```\n{route_status}```\n\n"
                except Exception:
                    response += "### Routes\n\nNo routes found\n\n"

                if detailed:
                    response += "### Vault Status\n\n"
                    try:
                        vault_status, _ = run_shell(f"oc exec -n {namespace} {release_name}-0 -- vault status")
                        response += f"```\n{vault_status}```\n\n"
                    except Exception:
                        response += "Unable to check vault status\n\n"

                response += "### Helm Release\n\n"
                try:
                    helm_status, _ = run_shell(f"helm list -n {namespace} -f {release_name}")
                    response += f"```\n{helm_status}```\n\n"
                except Exception:
                    response += "Unable to retrieve helm release info\n\n"

        response += "## Health Summary\n\n"
        response += _generate_health_summary(namespace, release_name)

        return text_response(response)
    except Exception as error:  # noqa: BLE001
        return text_response(f"Status check failed: {error}", is_error=True)


def _generate_health_summary(namespace: str | None, release_name: str | None) -> str:
    if not namespace:
        return "Cluster is accessible\nSpecify namespace for detailed health check\n"

    try:
        selector = f"-l app.kubernetes.io/instance={release_name}" if release_name else ""
        stdout, _ = run_shell(f"oc get pods -n {namespace} {selector} -o json")
        pods = json.loads(stdout)
        items = pods.get("items", [])

        if not items:
            return "No pods found\n"

        total = len(items)
        running = len([p for p in items if p.get("status", {}).get("phase") == "Running"])

        def is_ready(pod: dict[str, Any]) -> bool:
            return any(
                cond.get("type") == "Ready" and cond.get("status") == "True"
                for cond in pod.get("status", {}).get("conditions", [])
            )

        ready = len([p for p in items if is_ready(p)])

        summary = f"**Pods:** {running}/{total} running, {ready}/{total} ready\n"
        if running == total and ready == total:
            summary += "All pods are healthy\n"
        else:
            summary += "Some pods are not healthy\n"
        return summary
    except Exception:
        return "Unable to generate health summary\n"
